oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/NimbusJwtDecoder.java

```diff
@@ -309,6 +328,33 @@ private static URL toURL(String url) {
 			}
 		}
 
+		private static class CacheJWKSetCache implements JWKSetCache {
```

Comment:
Taking another look at this, the way the Nimbus cache refreshes can lead to more than one refresh call:

```java
JWKSet jwkSet = jwkSetCache.get();
if (jwkSetCache.requiresRefresh() || jwkSet == null) {
    // ... do refresh
    jwkSetCache.put(jwkSet);
}
```

If many threads run the above `if` statement at the same time, they will all induce calls to the authorization server.

I think it would be better to wrap a cache around a `ResourceRetreiver` implementation and use `Cache#get(key, callable)` since that provides synchronization guarantees. Then, you'd create a no-op `JWKSetCache` that never caches anything.